app/routes/_index.tsx

```diff
@@ -203,21 +216,93 @@ export default function Index() {
     setTimeout(() => setCopied(false), 2000);
   };
 
-  const [addForm, setAddForm] = React.useState<SteamappUpsert>({
-    app_id: 0,
-    base_image: "docker.io/library/debian:stable-slim",
-    apt_packages: [],
-    launch_type: "",
-    platform_type: "linux",
-    execs: [],
-    entrypoint: [],
-    cmd: [],
-    branch: "public",
-    beta_password: "",
-  });
+  const [addForm, setAddForm] = React.useState<SteamappUpsert>(defaultAddForm);
 
   const [editForm, setEditForm] = React.useState(addForm);
 
+  // fragment
+  React.useEffect(() => {
+    const handleHashChange = () => {
+      const fragment = window.location.hash.slice(1);
+
+      if (fragment === "add") {
+        setActivity("adding");
+      } else if (fragment.startsWith("edit/")) {
+        const appId = parseInt(fragment.split('/')[1]);
+        if (!isNaN(appId) && appId >= 0) {
+          const steamappIndex = steamapps.findIndex(s => s.app_id === appId);
+          if (steamappIndex >= 0) {
+            getSteamappDetails(steamappIndex)
+              .then(() => setEditForm(steamapps[steamappIndex] as SteamappUpsert))
+              .then(() => setActivity('editing'))
+              .catch(handleErr);
+          }
+        }
+      } else if (fragment.startsWith("view/")) {
+        const appId = parseInt(fragment.split('/')[1]);
+        if (!isNaN(appId) && appId >= 0) {
+          const steamappIndex = steamapps.findIndex(s => s.app_id === appId);
+          if (steamappIndex >= 0) {
+            getSteamappDetails(steamappIndex)
+              .then(() => setViewingSteamappIndex(steamappIndex))
+              .then(() => setActivity('viewing'))
+              .catch(handleErr);
+          }
+        }
+      } else if (fragment === "") {
+        setActivity(undefined);
+      }
+    }
+
+    handleHashChange();
+
+    window.addEventListener('hashchange', handleHashChange);
+
+    return () => {
+      window.removeEventListener('hashchange', handleHashChange);
+    };
+  }, [steamapps, getSteamappDetails, handleErr])
+
+  const setActivityWithFragment = (newActivity: typeof activity, appId?: number) => {
+    if (newActivity === 'adding') {
+      window.location.hash = '#add';
+    } else if (newActivity === 'editing' && appId) {
+      window.location.hash = `#edit/${appId}`;
+    } else if (newActivity === 'viewing' && appId) {
+      window.location.hash = `#view/${appId}`;
+    } else {
+      window.location.hash = '';
+    }
+    setActivity(newActivity);
+  };
```

Comment:
I think it would be better if this function _just_ set the activity and then a `useEffect` reacted to that state change to update the fragment.